kernel/src/loader/mod.rs

```diff
@@ -144,6 +148,87 @@ where
     Ok(GuestAddress(ehdr.e_entry as usize))
 }
 
+#[cfg(target_arch = "aarch64")]
+pub fn load_kernel<F>(
+    guest_mem: &GuestMemory,
+    kernel_image: &mut F,
+    start_address: usize,
+) -> Result<GuestAddress>
+where
+    F: Read + Seek,
+{
+    /* Kernel boot protocol is specified in the kernel docs
+    Documentation/arm/Booting and Documentation/arm64/booting.txt.
+
+    ======aarch64 kernel header========
+    u32 code0;			/* Executable code */
+    u32 code1;			/* Executable code */
+    u64 text_offset;		/* Image load offset, little endian */
+    u64 image_size;		/* Effective Image size, little endian */
+    u64 flags;			/* kernel flags, little endian */
+    u64 res2	= 0;		/* reserved */
+    u64 res3	= 0;		/* reserved */
+    u64 res4	= 0;		/* reserved */
+    u32 magic	= 0x644d5241;	/* Magic number, little endian, "ARM\x64" */
+    u32 res5;			/* reserved (used for PE COFF offset) */
+    ====================================
+     */
+    const AARCH64_KERNEL_LOAD_ADDR: usize = 0x80000;
+    const AARCH64_MAGIC_NUMBER: u32 = 0x644d5241;
+    let mut kernel_load_offset = AARCH64_KERNEL_LOAD_ADDR;
+
+    /* Look for the magic number inside the elf header. */
+    let magic_offset_header = 2 * mem::size_of::<u32>() + 6 * mem::size_of::<u64>(); // This should total 56.
+
+    kernel_image
+        .seek(SeekFrom::Start(magic_offset_header as u64))
+        .map_err(|_| Error::SeekKernelImage)?;
+    let mut magic_number: u32 = 0;
+    unsafe {
+        sys_util::read_struct(kernel_image, &mut magic_number)
```

Comment:
It just crossed my mind, I think it might make sense to define a `kernel_header` structure and read all the kernel header to it. The advantage would be that you don't have to seek through the file to read individual header fields. It also improves code readability. On the downside you keep in memory more fields than you actually need (not sure if this is actually a disadvantage as we drop the structure after exiting from this function.

Specifically, what I mean is the following (I didn't check if it compiles):
```rust
struct kernel_header {
  u32 code0,			/* Executable code */
  u32 code1,			/* Executable code */
  u64 text_offset,		/* Image load offset, little endian */
  u64 image_size,		/* Effective Image size, little endian */
  u64 flags,			/* kernel flags, little endian */
  u64 res2, 		/* reserved */
  u64 res3,		/* reserved */
  u64 res4,		/* reserved */
  u32 magic,	/* Magic number, little endian, "ARM\x64" */
  u32 res5,			/* reserved (used for PE COFF offset) */
}

....

let mut elf_kernel_header: kernel_header = unsafe { mem::zeroed() };
sys_util::read_struct(kernel_image, &mut elf_kernel_header).map_err(|_| Error::ReadProgramHeader)?;

// Check magic
if elf_kernel_header.magic != AARCH64_MAGIC_NUMBER {
        return Err(Error::InvalidElfMagicNumber);
}

// We might need extra checks here
 let kernel_load_offset = elf_kernel_header.text_offset + start_address;
```

Reply:
I added the comment that contains the whole header exactly to make it more readable and easy to follow. Defining a structure with 10 fields just to be using 3 out of them does not add any value in my opinion.